### PR TITLE
Add the procedue of FIN STREAM frame

### PR DESCRIPTION
**Pull request by scw00**

### CODE REVIEW

File: iocore/net/quic/QUICStream.cc [hunk 1]
```diff
@@ -248,9 +248,25 @@ QUICStream::_signal_write_event(bool direct)
   }
 }
 
-void
+QUICError
 QUICStream::_write_to_read_vio(const std::shared_ptr<const QUICStreamFrame> &frame)
 {
+  if (frame->has_fin_flag()) {
+    _closed = true;
```
Comment (maskit) [inline iocore/net/quic/QUICStream.cc]: I'm not a big fun of this flag. Can we use QUICStreamState insted?

Comment (scw00) [reply iocore/net/quic/QUICStream.cc]: Updated later!

**Comment by scw00** [comment]
@maskit Updated！Can you please review again !

File: iocore/net/quic/QUICStream.cc [hunk 2]
```diff
@@ -409,3 +417,63 @@ QUICStream::largest_offset_sent()
...
+
+QUICError
+QUICStream::_check_and_set_fin_offset(QUICOffset offset, bool set_last_offset)
+{
+  if (set_last_offset) {
+    if (!this->is_valid_fin_offset(offset)) {
+      return QUICError(QUICErrorClass::QUIC_TRANSPORT, QUICErrorCode::FINAL_OFFSET_ERROR);
+    }
+
+    this->_last_offset = offset;
```
Comment (masaori335) [inline iocore/net/quic/QUICStream.cc]: Do you mean final offset by `_last_offset`? Why don't you call it `_final_offset`?

File: iocore/net/quic/QUICStream.cc [hunk 3]
```diff
@@ -409,3 +417,63 @@ QUICStream::largest_offset_sent()
...
+
+  if (this->_last_offset != last_offset) {
+    return false;
+  }
+
+  return true;
+}
+
+QUICErrorCode
+QUICStream::get_error_code() const
```
Comment (masaori335) [inline iocore/net/quic/QUICStream.cc]: Where is this method called?

File: iocore/net/quic/QUICStream.cc [hunk 4]
```diff
@@ -306,16 +316,14 @@ QUICStream::recv(const std::shared_ptr<const QUICStreamFrame> frame)
   if (this->_recv_offset > frame->offset()) {
     // Do nothing. Just ignore STREAM frame.
     return QUICError(QUICErrorClass::NONE);
-  } else if (this->_recv_offset == frame->offset()) {
```
Comment (masaori335) [inline iocore/net/quic/QUICStream.cc]: What is the advantage of inserting every frames in `_received_stream_frame_buffer` once?

Comment (scw00) [reply iocore/net/quic/QUICStream.cc]: I'd like to leave the received buffer independently. That means we don't need to care about the dup frame, disorder frame or other one. Just push it in buffer and do the logic in received buffer.  Return error code if push failed. Then we can always get the frame in order.

File: iocore/net/quic/QUICStream.cc (hunk 3 above)
Comment (scw00) [reply iocore/net/quic/QUICStream.cc]: It seems we missing rst error msg. And we should reserve this message sometimes.

File: iocore/net/quic/QUICStream.cc (hunk 2 above)
Comment (scw00) [reply iocore/net/quic/QUICStream.cc]: That's right, I'll fix it.

File: iocore/net/quic/QUICStream.cc (hunk 4 above)
Comment (masaori335) [reply iocore/net/quic/QUICStream.cc]: Got it. My concern was that when `_received_stream_frame_buffer` become large, looking up the frame could be expensive. But we can optimize it later if we need.

Comment (scw00) [reply iocore/net/quic/QUICStream.cc]: That's right !

**Comment by scw00** [comment]
@masaori335 updated please review again !

File: iocore/net/quic/QUICStream.cc (hunk 3 above)
Comment (maskit) [reply iocore/net/quic/QUICStream.cc]: I'm not sure how we can use it yet. Let's focus on FIN handling on this PR.

File: iocore/net/quic/test/test_QUICStreamManager.cc [hunk 5]
```diff
@@ -165,3 +165,85 @@ TEST_CASE("QUICStreamManager_total_offset_sent", "[quic]")
...
+  app_map.set_default(&mock_app);
+  QUICStreamManager sm(&tx, &app_map);
+  std::shared_ptr<QUICTransportParameters> local_tp = std::make_shared<QUICTransportParametersInEncryptedExtensions>();
+  local_tp->add(QUICTransportParameterId::INITIAL_MAX_STREAM_DATA,
+                std::unique_ptr<QUICTransportParameterValue>(new QUICTransportParameterValue(4096, 4)));
+  std::shared_ptr<QUICTransportParameters> remote_tp =
+    std::make_shared<QUICTransportParametersInClientHello>(static_cast<QUICVersion>(0), static_cast<QUICVersion>(0));
+
+  sm.init_flow_control_params(local_tp, remote_tp);
+  char data[1024] = {0};
```
Comment (maskit) [inline iocore/net/quic/test/test_QUICStreamManager.cc]: Let's make this `uint8_t data[1024]`, then we can remove the `reinterpret_cast`s.

**Comment by scw00** [comment]
I am not satisfied with this pr, because it did a lot of repetitive work. closes it for now !

**Pull request closed without merging**